### ISSUE TITLE
[Request] Ability to reset a channel to beginning of file

### ISSUE THREAD

**Issue by LouisJenkinsCS** [chapel-lang/chapel#13204]
I've asked on gitter without any answer, and I have looked around for some kind of `reset` method to call on a `channel`. `mark`/`revert` seem a bit like overkill. Previously I was creating a new `reader` each time to get back to the beginning of a file again via `f.reader()`. However, I had no idea until now how expensive that was.

**file.reader**

https://github.com/chapel-lang/chapel/blob/034cbfa5133ef66526a7e5ab741ad2423d1438f0/modules/standard/IO.chpl#L2443-L2456

**channel.init**

https://github.com/chapel-lang/chapel/blob/034cbfa5133ef66526a7e5ab741ad2423d1438f0/modules/standard/IO.chpl#L1865-L1878

**qio_channel_create**

https://github.com/chapel-lang/chapel/blob/1c73789d1591b4f1de497a653c70406660fb943e/runtime/src/qio/qio.c#L1642-L1661

**_qio_channel_init_file**

https://github.com/chapel-lang/chapel/blob/1c73789d1591b4f1de497a653c70406660fb943e/runtime/src/qio/qio.c#L1588-L1640

**_qio_channel_init_file_internal**

https://github.com/chapel-lang/chapel/blob/1c73789d1591b4f1de497a653c70406660fb943e/runtime/src/qio/qio.c#L1452-L1514

Why did I reference all of this code? Its the surface of what each call to `f.reader()` does. Hence, I would like to avoid this overhead at all costs. In my application, I have a custom binary format that has offsets encoded in the header of the file, which we then use to advance the reader ahead to that offset to read in the data. I do this in parallel, across each task per locale (one task per core per locale). Is there a more minimal and efficient way to get back to the beginning of a file other than `mark`/`revert`? If not, can I request that one be created?

**Comment by bradcray** [comment]
I think your TIO example doesn't work because of the following clause in `mark()`'s documentation:

This function can only be called on a channel with ``locking==false``.  So I think the fix is [this](https://tio.run/##HcpBDoMgEAXQfU9BXNHE4AEaDjORjyWOQL6WHh9pd2/x1rdUaO9NaKLxplRkOy0Bbckf1Wk2qRwlwPH7fD1@i2NFR0gArZZ1T3nzUfTECHSHcLd/EQ28hnu/AQ).

[Caveat: I did not read all of the code you pasted above, so may have missed something crucial]

**Comment by LouisJenkinsCS** [comment]
Why isn't `locking` disabled by default?

**Comment by LouisJenkinsCS** [comment]
Anyway I'd also like to keep this issue open as I would like an explicit `channel.reset`, even if it uses `mark`/`revert` under the hood, or does something smarter.

**Comment by bradcray** [comment]
> Why isn't locking disabled by default?

I believe this is an example of Chapel's typical desire to provide safety in the context of parallelism first and a way to enable performance second.

> Anyway I'd also like to keep this issue open as I would like an explicit channel.reset, even if it uses mark/revert under the hood, or does something smarter.

I'd suggest rewriting your issue to make it sound more like a feature request for channel.reset (which I think is a reasonable request) and less like... I don't even know what to call it.

**Comment by LouisJenkinsCS** [comment]
Getting a sense for overhead of calls to `mark`/`revert`...

**channel.mark**

https://github.com/chapel-lang/chapel/blob/034cbfa5133ef66526a7e5ab741ad2423d1438f0/modules/standard/IO.chpl#L2158-L2160

**qio_channel_mark**

https://github.com/chapel-lang/chapel/blob/1c73789d1591b4f1de497a653c70406660fb943e/runtime/src/qio/qio.c#L3577-L3580

**qio_channel_mark_maybe_flush_bits**

https://github.com/chapel-lang/chapel/blob/1c73789d1591b4f1de497a653c70406660fb943e/runtime/src/qio/qio.c#L3499-L3575

**_qio_flush_bits_if_needed_unlocked**

https://github.com/chapel-lang/chapel/blob/1c73789d1591b4f1de497a653c70406660fb943e/runtime/src/qio/qio.c#L2372-L2423


And more, including for `revert`.

> d suggest rewriting your issue to make it sound more like a feature request for channel.reset (which I think is a reasonable request) and less like... I don't even know what to call it.

I'll remove the part where I say that `mark`/`revert` is buggy.

**Comment by LouisJenkinsCS** [comment]
Okay so `mark`/`revert` is _significantly_ faster than creating a new reader (at least 2 orders of magnitude).

**Comment by mppf** [comment]
> In my application, I have a custom binary format that has offsets encoded in the header of the file, which we then use to advance the reader ahead to that offset to read in the data.... Is there a more minimal and efficient way to get back to the beginning of a file other than mark/revert? 

I'm glad that mark/revert is working for you. In your particular case I think it'd be reasonable to use multiple channels. One channel uses mark/revert and you might consider it a benefit if it keeps buffered the file header. You'd use this channel to read offsets from the header. Then, to read the data from that offset, you'd make a new channel.

AFAIK most of the overhead in creating a new channel is actually in doing some memory allocations. I think it'd be interesting to consider ways to reuse the memory allocations. If you are mmap'ing the file, then creating the channel should not have much overhead beyond some memory allocation.

Lastly, if you're reading a really small amount (say 1 word), you might be happier mmap'ing the whole file. We've in the past considered enabling arrays to be backed by a file. Would having an array view of your file make this code easier to write?

Anyway, the reason that files do not have a "seek" method is to avoid race conditions among many tasks using the file. Channels don't have a "seek" method also because it's easier to write race conditions with that idiom and because the I/O code assumes that the offsets to read/write are monotonically increasing (as a simplifying assumption). It's important in the design of the I/O system that channels can buffer up data.

I do think it'd be interesting to consider enabling a channel to "seek" to an earlier offset. But I'd prefer to first create a facility for mmap'ing a file into an array, if that idiom makes sense for this use case.

**Comment by LouisJenkinsCS** [comment]
> We've in the past considered enabling arrays to be backed by a file. Would having an array view of your file make this code easier to write?

Most definitely! So long as the array  manages how many pages are mapped into memory at once so as to not hit OOM. I am a bit foggy on the low level details of mmap. At least the file-backed array would make it more intuitive to write the code.

**Comment by mppf** [comment]
> So long as the array manages how many pages are mapped into memory at once so as to not hit OOM.

That is something the OS normally handles.

Perhaps we could discuss the design of this further on another issue?

**Comment by LouisJenkinsCS** [comment]
Sure, I suppose I'll defer the discussion until later. However for now, is there a way to manipulate qio internals to just reset the offset back to 0 without going through `mark`/`revert`? Even if it is 'unsafe'.

**Comment by mppf** [comment]
> However for now, is there a way to manipulate qio internals to just reset the offset back to 0 without going through mark/revert? Even if it is 'unsafe'.

That might be possible, but it might actually be hard, because as I said the I/O system was built with the simplifying assumption that channel offsets only ever increase. If I were going to try doing it, I'd try:
 * dropping all buffers in the channel's buffer object
 * changing the position (stored at the top of the "mark stack")
 * hoping that I/O at the new position to re-create the buffers
